File: primitives/tool_use/code/python/permissions.py
"""
Tool Permissions — Always / Ask / Never gating for tool execution.

Each registered tool carries a permission tier. The gate consults it before any
tool runs: ALWAYS executes silently, ASK routes to a human-approval callback
(the human-in-the-loop seam), and NEVER refuses outright. This is the hard
boundary that stops an agent invoking a destructive tool without a human in the
loop — the permission model the production-agent literature treats as
load-bearing. Unclassified tools default to ASK: fail safe, never silently run.

Design doc:  ../../design.md
Overview:    ../../overview.md
"""

from __future__ import annotations

from dataclasses import dataclass
from enum import Enum
from typing import Protocol

from primitives.tool_use.schemas.state import ToolCall

# ── Core types ────────────────────────────────────────────────────────────────


class Permission(str, Enum):
    """How a tool may be invoked."""

    ALWAYS = "always"  # run without asking
    ASK = "ask"  # require human approval first
    NEVER = "never"  # refuse outright


@dataclass
class PermissionDecision:
    """The outcome of gating one tool call."""

    allowed: bool
    permission: Permission
    reason: str = ""


# ── Interfaces ────────────────────────────────────────────────────────────────


class ApprovalPrompt(Protocol):
    """Asks a human to approve one ASK-tier tool call; returns True to allow."""

    def __call__(self, call: ToolCall, reason: str) -> bool: ...


# ── Implementation ────────────────────────────────────────────────────────────


class PermissionGate:
    """Maps tool name → :class:`Permission` and decides whether a call may run.

    A tool absent from the map falls back to ``default`` (ASK) — a tool nobody
    classified is treated as needing approval, never silently allowed.
    """

    def __init__(
        self,
        permissions: dict[str, Permission],
        *,
        approve: ApprovalPrompt | None = None,
        default: Permission = Permission.ASK,
    ) -> None:
        self._permissions = dict(permissions)
        self._approve = approve
        self._default = default

    def permission_for(self, tool: str) -> Permission:
        return self._permissions.get(tool, self._default)

    def check(self, call: ToolCall) -> PermissionDecision:
        permission = self.permission_for(call.tool)
        if permission is Permission.ALWAYS:
            return PermissionDecision(True, permission, "allowed (ALWAYS)")
        if permission is Permission.NEVER:
            return PermissionDecision(False, permission, f"tool {call.tool!r} denied (NEVER)")
        reason = f"tool {call.tool!r} requires approval"
        if self._approve is None:
            return PermissionDecision(False, permission, f"{reason} but no approver configured")
        approved = self._approve(call, reason)
        verdict = "approved" if approved else "declined"
        return PermissionDecision(approved, permission, f"{reason} — {verdict}")


# ── Example ───────────────────────────────────────────────────────────────────

if __name__ == "__main__":
    gate = PermissionGate(
        {
            "get_weather": Permission.ALWAYS,
            "delete_file": Permission.NEVER,
            "send_email": Permission.ASK,
        },
        approve=lambda call, reason: call.args.get("to") == "team@example.com",
    )
    samples = [
        ("get_weather", {"city": "Tokyo"}),
        ("delete_file", {"path": "/etc/passwd"}),
        ("send_email", {"to": "team@example.com"}),
        ("send_email", {"to": "stranger@example.net"}),
    ]
    for name, args in samples:
        decision = gate.check(ToolCall(tool=name, args=args))
        print(f"{name}: allowed={decision.allowed} — {decision.reason}")

File: primitives/tool_use/code/typescript/vercel-ai-sdk/permissions.ts
/**
 * Tool Permissions — Always / Ask / Never gating for tool execution.
 *
 * Pattern: Tool Use (the permission tier consulted before every dispatch).
 * Framework-agnostic: pure TypeScript, no SDK dependency — it slots under the
 *   Vercel AI SDK dispatcher in tool-use.ts, or any other tool loop. Mirrors the
 *   Python sibling ../../python/permissions.py.
 * Idioms: ALWAYS runs silently, ASK routes to a human-approval callback (the
 *   human-in-the-loop seam), NEVER refuses outright. A tool absent from the map
 *   defaults to ASK — fail safe, never silently run.
 * Design doc: ../../../design.md
 *
 * Run:  npx tsx permissions.ts   (zero runtime deps — nothing to install)
 *
 * Note: ESM only (matches the canonical TypeScript project layout).
 * tsconfig snippet: { "module": "ESNext", "moduleResolution": "Bundler", "target": "ES2022" }
 */

// ── Core types ────────────────────────────────────────────────────────────────

/** How a tool may be invoked. Mirrors the Python `Permission` str-enum. */
const Permission = {
  ALWAYS: "always", // run without asking
  ASK: "ask", // require human approval first
  NEVER: "never", // refuse outright
} as const;
type Permission = (typeof Permission)[keyof typeof Permission];

/** One tool invocation request from the model. Mirrors schemas/state.py. */
interface ToolCall {
  tool: string;
  args: Record<string, unknown>;
  id?: string;
}

/** The outcome of gating one tool call. */
interface PermissionDecision {
  allowed: boolean;
  permission: Permission;
  reason: string;
}

/** Asks a human to approve one ASK-tier tool call; returns true to allow. */
type ApprovalPrompt = (call: ToolCall, reason: string) => boolean;

// ── Gate ──────────────────────────────────────────────────────────────────────

/**
 * Maps tool name -> {@link Permission} and decides whether a call may run.
 *
 * A tool absent from the map falls back to `default` (ASK) — a tool nobody
 * classified is treated as needing approval, never silently allowed.
 */
class PermissionGate {
  private readonly permissions: Map<string, Permission>;
  private readonly approve?: ApprovalPrompt;
  private readonly fallback: Permission;

  constructor(
    permissions: Record<string, Permission>,
    opts: { approve?: ApprovalPrompt; default?: Permission } = {},
  ) {
    this.permissions = new Map(Object.entries(permissions));
    this.approve = opts.approve;
    this.fallback = opts.default ?? Permission.ASK;
  }

  permissionFor(tool: string): Permission {
    return this.permissions.get(tool) ?? this.fallback;
  }

  check(call: ToolCall): PermissionDecision {
    const permission = this.permissionFor(call.tool);
    if (permission === Permission.ALWAYS) {
      return { allowed: true, permission, reason: "allowed (ALWAYS)" };
    }
    if (permission === Permission.NEVER) {
      return { allowed: false, permission, reason: `tool '${call.tool}' denied (NEVER)` };
    }
    const reason = `tool '${call.tool}' requires approval`;
    if (this.approve === undefined) {
      return { allowed: false, permission, reason: `${reason} but no approver configured` };
    }
    const approved = this.approve(call, reason);
    const verdict = approved ? "approved" : "declined";
    return { allowed: approved, permission, reason: `${reason} — ${verdict}` };
  }
}

// ── Example ────────────────────────────────────────────────────────────────────

function main(): void {
  const gate = new PermissionGate(
    {
      get_weather: Permission.ALWAYS,
      delete_file: Permission.NEVER,
      send_email: Permission.ASK,
    },
    { approve: (call) => call.args.to === "team@example.com" },
  );

  const samples: ToolCall[] = [
    { tool: "get_weather", args: { city: "Tokyo" } },
    { tool: "delete_file", args: { path: "/etc/passwd" } },
    { tool: "send_email", args: { to: "team@example.com" } },
    { tool: "send_email", args: { to: "stranger@example.net" } },
  ];
  for (const call of samples) {
    const decision = gate.check(call);
    console.log(`${call.tool}: allowed=${decision.allowed} — ${decision.reason}`);
  }
}

// ESM entry-point check.
if (import.meta.url === `file://${process.argv[1]}`) {
  main();
}

export { Permission, PermissionGate };
export type { ToolCall, PermissionDecision, ApprovalPrompt };
